src/debug_wait.c

```diff
@@ -102,81 +210,32 @@ ts_debug_waitpoint_wait(DebugWait *waitpoint, bool blocking)
+	lock_release_result = LockRelease(&point.tag, ShareLock, true);
 	Assert(lock_release_result);
 
-	ereport(DEBUG3, (errmsg("proceeding after waitpoint '%s'", waitpoint->tagname)));
-}
-
-static void
-debug_waitpoint_enable(DebugWait *waitpoint)
-{
-	ereport(DEBUG1, (errmsg("enabling waitpoint \"%s\"", waitpoint->tagname)));
-	if (LockAcquire(&waitpoint->tag, ExclusiveLock, true, true) == LOCKACQUIRE_NOT_AVAIL)
-		ereport(NOTICE, (errmsg("debug waitpoint \"%s\" already enabled", waitpoint->tagname)));
-}
-
-static void
-debug_waitpoint_release(DebugWait *waitpoint)
-{
-	ereport(DEBUG1, (errmsg("releasing waitpoint \"%s\"", waitpoint->tagname)));
-	if (!LockRelease(&waitpoint->tag, ExclusiveLock, true))
-		elog(ERROR, "cannot release waitpoint \"%s\"", waitpoint->tagname);
+	ereport(DEBUG3, (errmsg("proceeding after debug point '%s'", point.name)));
 }
 
 /*
- * Enable a waitpoint to block when being reached.
- *
- * This function will always succeed since we will not lock the waitpoint if
- * it is already locked. A notice will be printed if the waitpoint is already
- * enabled.
+ * Produce an error in case if the debug point is enabled.
  */
-TS_FUNCTION_INFO_V1(ts_debug_waitpoint_enable);
-
-Datum
-ts_debug_waitpoint_enable(PG_FUNCTION_ARGS)
-{
-	text *tag = PG_GETARG_TEXT_PP(0);
-	DebugWait waitpoint;
-
-	if (PG_ARGISNULL(0))
-		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE), errmsg("no tag provided")));
-
-	ts_debug_waitpoint_init(&waitpoint, text_to_cstring(tag));
-	debug_waitpoint_enable(&waitpoint);
-	PG_RETURN_VOID();
-}
-
-/*
- * Release a waitpoint allowing execution to proceed.
- */
-TS_FUNCTION_INFO_V1(ts_debug_waitpoint_release);
-Datum
-ts_debug_waitpoint_release(PG_FUNCTION_ARGS)
+void
+ts_debug_point_error_injection(const char *name)
 {
-	text *tag = PG_GETARG_TEXT_PP(0);
-	DebugWait waitpoint;
+	DebugPoint point;
+	LockAcquireResult lock_acquire_result;
 
-	if (PG_ARGISNULL(0))
-		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE), errmsg("no tag provided")));
+	debug_point_init(&point, name);
 
-	ts_debug_waitpoint_init(&waitpoint, text_to_cstring(tag));
-	debug_waitpoint_release(&waitpoint);
-	PG_RETURN_VOID();
-}
-
-/*
- * Get the waitpoint identifier from the tag name.
- */
-TS_FUNCTION_INFO_V1(ts_debug_waitpoint_id);
-
-Datum
-ts_debug_waitpoint_id(PG_FUNCTION_ARGS)
-{
-	text *tag = PG_GETARG_TEXT_PP(0);
+	lock_acquire_result = LockAcquire(&point.tag, ExclusiveLock, true, true);
+	if (lock_acquire_result != LOCKACQUIRE_NOT_AVAIL)
+	{
+		/* release/decrement lock count for OK, ALREADY_HELD cases */
+		LockRelease(&point.tag, ExclusiveLock, true);
+	}
 
-	if (PG_ARGISNULL(0))
-		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE), errmsg("no tag provided")));
+	if (lock_acquire_result == LOCKACQUIRE_OK)
+		return;
 
-	PG_RETURN_UINT64(ts_debug_waitpoint_tag_to_id(text_to_cstring(tag)));
+	ereport(ERROR, (errmsg("debug point '%s' error injection", point.name)));
```

Comment:
Do we really want to throw the same error for the `LOCKACQUIRE_ALREADY_CLEAR` case?

Reply:
Looks like we should, it still takes the lock and do reference counter increment under the hood